pkg/pipeline/ingest/metrics.go

```diff
@@ -6,21 +6,69 @@ import (
 	"github.com/prometheus/client_golang/prometheus"
 )
 
+var (
+	latencyHistogram = operational.DefineMetric(
+		"ingest_latency_ms",
+		"Latency between flow end time and ingest time, in milliseconds",
+		operational.TypeHistogram,
+		"stage",
+	)
+	batchSizeSummary = operational.DefineMetric(
+		"ingest_flows_processed", // This is intentionally named to emphasize its utility for flows counting, despite being a batch size distribution
+		"Provides number of flows processed, batches processed, and batch size stats (in number of flows)",
+		operational.TypeSummary,
+		"stage",
+	)
+	batchSizeBytesSummary = operational.DefineMetric(
+		"ingest_batch_size_bytes",
+		"Ingested batch size distribution, in bytes",
+		operational.TypeSummary,
+		"stage",
+	)
+	errorsCounter = operational.DefineMetric(
+		"ingest_errors",
+		"Counter of errors during ingestion",
+		operational.TypeCounter,
+		"stage", "type", "code",
+	)
+)
+
 type metrics struct {
 	*operational.Metrics
 	stage          string
-	flowsProcessed prometheus.Counter
+	stageType      string
+	stageDuration  prometheus.Observer
+	latency        prometheus.Histogram
+	batchSize      prometheus.Summary
+	batchSizeBytes prometheus.Summary
+	errors         *prometheus.CounterVec
 }
 
-func newMetrics(opMetrics *operational.Metrics, stage string, inGaugeFunc func() int) *metrics {
+func newMetrics(opMetrics *operational.Metrics, stage, stageType string, inGaugeFunc func() int) *metrics {
 	opMetrics.CreateInQueueSizeGauge(stage, inGaugeFunc)
 	return &metrics{
 		Metrics:        opMetrics,
 		stage:          stage,
-		flowsProcessed: opMetrics.CreateFlowsProcessedCounter(stage),
+		stageType:      stageType,
+		latency:        opMetrics.NewHistogram(&latencyHistogram, []float64{.001, .01, .1, 1, 10, 100, 1000, 10000}, stage),
+		stageDuration:  opMetrics.GetOrCreateStageDurationHisto().WithLabelValues(stage),
+		batchSize:      opMetrics.NewSummary(&batchSizeSummary, stage),
+		batchSizeBytes: opMetrics.NewSummary(&batchSizeBytesSummary, stage),
+		errors:         opMetrics.NewCounterVec(&errorsCounter),
```

Comment:
What's the thumb rule for choosing Summary over Histogram?
According to [Histograms and summaries](https://prometheus.io/docs/practices/histograms/), considering client performance, summaries are more expensive:
```
Observations are expensive due to the streaming quantile calculation.
```

Reply:
I don't know if there's a rule of thumb, I tend to prefer histograms also because they provide more flexibility in queries.
On the flip side, summaries tend to generate less metrics (although it depends on configuration, but that's generally speaking) which is also something the prometheus server appreciates.
But TBH, here I just didn't change their types: goflow2 uses summaries which, previously, were reused in ingest_grpc.